File: Sources/HedosCLI/Commands/Speak.swift
import ArgumentParser
import Foundation
import HedosKernel

struct Speak: AsyncParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "speak",
        abstract: "Generate speech from text and write it to a .wav file.")

    @OptionGroup var global: GlobalOptions

    @Argument(help: "Model id or name.")
    var model: String

    @Argument(help: "The text to speak.")
    var text: String

    @Option(name: .long, help: "Voice to use (defaults to the model's first voice).")
    var voice: String?

    @Option(name: .long, help: "Speaking speed multiplier.")
    var speed: Double = 1.0

    @Option(name: [.customShort("o"), .long], help: "Output file path (.wav).")
    var output: String?

    func run() async throws {
        let kernel = Session.kernel()
        let shelf = try await Session.shelf(kernel)
        let record = try Session.resolve(model, in: shelf, capability: .speak)

        let voices = (try? await kernel.voices(for: record.id)) ?? []
        let chosenVoice = voice ?? voices.first ?? "default"

        let payload: JSONValue = .object([
            "text": .string(text),
            "voice": .string(chosenVoice),
            "speed": .double(speed),
        ])

        var pcm = Data()
        var sampleRate = 24000
        for try await chunk in try await kernel.invoke(record.id, .speak, payload: payload) {
            switch chunk {
            case .audio(let frame):
                pcm.append(frame.data)
                sampleRate = frame.sampleRate
            case .status(let status):
                if !global.json { Out.err(status) }
            default:
                break
            }
        }
        guard !pcm.isEmpty else { throw CLIError("\(record.displayName) produced no audio.") }

        let artifact = try await kernel.saveSpeech(
            modelID: record.id, voice: chosenVoice, text: text,
            speed: speed, sampleRate: sampleRate, pcm: pcm)
        let stored = try await kernel.artifactStore.url(id: artifact.id)
        let finalPath = try deliver(stored, to: output)

        if global.json {
            try Out.json(MediaReport(
                model: record.id, path: finalPath, voice: chosenVoice,
                durationMs: artifact.durationMs))
        } else {
            Out.line(finalPath)
        }
    }
}

func deliver(_ stored: URL?, to output: String?) throws -> String {
    guard let stored else { throw CLIError("the output was not written to disk.") }
    guard let output else { return stored.path }
    let destination = URL(fileURLWithPath: (output as NSString).expandingTildeInPath)
    var isDirectory: ObjCBool = false
    if FileManager.default.fileExists(atPath: destination.path, isDirectory: &isDirectory),
        isDirectory.boolValue
    {
        throw CLIError("\(destination.path) is a directory — pass a file path to -o.")
    }
    try? FileManager.default.removeItem(at: destination)
    try FileManager.default.copyItem(at: stored, to: destination)
    return destination.path
}

struct MediaReport: Encodable {
    let model: String
    let path: String
    var voice: String? = nil
    var durationMs: Int? = nil
}

File: cli/src/commands/speak.rs
//! `hedos speak <model> <text>` — synthesize speech and write a WAV file.

use std::collections::BTreeMap;
use std::path::PathBuf;

use clap::Args;
use kernel::capabilities::CapabilityChunk;
use kernel::records::{Capability, JsonValue};
use runtime::sidecar::DEFAULT_SAMPLE_RATE;

use crate::error::CliError;
use crate::support::output::Out;
use crate::support::session::{self, Session};

/// Arguments for `speak`.
#[derive(Args)]
pub struct SpeakArgs {
    /// The speech model (name, alias, or id).
    model: String,
    /// The text to speak.
    text: String,
    /// The voice to use (default: the model's first bundled voice).
    #[arg(long)]
    voice: Option<String>,
    /// Playback speed multiplier.
    #[arg(long, default_value_t = 1.0)]
    speed: f64,
    /// Where to write the `.wav` (default: `<model>.wav` in the current dir).
    #[arg(short, long)]
    output: Option<PathBuf>,
}

/// Run the `speak` command.
pub async fn run(args: SpeakArgs, out: &Out) -> Result<(), CliError> {
    let session = Session::open()?;
    let shelf = session.shelf_or_discover().await?;
    let record = session::resolve(&args.model, &shelf, Some(&Capability::speak()))?;

    let voice = match args.voice {
        Some(voice) => Some(voice),
        None => session.kernel.voices(&record.id).await?.into_iter().next(),
    };
    let payload = speak_payload(&args.text, voice.as_deref(), args.speed);
    let mut stream = session
        .kernel
        .invoke(&record.id, Capability::speak(), payload)
        .await?;

    let mut pcm: Vec<u8> = Vec::new();
    let mut sample_rate = DEFAULT_SAMPLE_RATE;
    while let Some(result) = stream.recv().await {
        match result? {
            CapabilityChunk::Audio(frame) => {
                if pcm.is_empty() {
                    sample_rate = frame.sample_rate;
                }
                pcm.extend_from_slice(&frame.data);
            }
            CapabilityChunk::Status(status) => out.err(&status),
            _ => {}
        }
    }
    if pcm.is_empty() {
        return Err(CliError::new(format!(
            "{} produced no audio",
            record.display_name()
        )));
    }

    let rate = u32::try_from(sample_rate).unwrap_or(DEFAULT_SAMPLE_RATE as u32);
    let wav = runtime::audio::wav_from_pcm(&pcm, rate);
    let path = args
        .output
        .unwrap_or_else(|| crate::support::paths::default_path(record.display_name(), "wav"));
    std::fs::write(&path, wav)?;

    out.line(&format!("wrote {}", path.display()));
    out.json(&serde_json::json!({
        "model": record.id,
        "path": path.display().to_string(),
        "voice": voice,
    }));
    Ok(())
}

/// A `speak` payload: text plus an optional voice and a speed multiplier.
fn speak_payload(text: &str, voice: Option<&str>, speed: f64) -> JsonValue {
    let mut payload = BTreeMap::new();
    payload.insert("text".to_owned(), JsonValue::String(text.to_owned()));
    if let Some(voice) = voice {
        payload.insert("voice".to_owned(), JsonValue::String(voice.to_owned()));
    }
    payload.insert("speed".to_owned(), JsonValue::Double(speed));
    JsonValue::Object(payload)
}
